Fix virtiofs xfstests

liuyd96 opened the pull request:
Before this commit

guest crash
```
04:23:06 DEBUG| Sending command: ./check -virtiofs -E blacklist
05:00:35 INFO | [qemu output] error: kvm run failed Bad address
05:00:35 INFO | [qemu output]  PC=ffff5eedb65cb5b0 X00=ffffa072a22c0000 X01=0000ffff909a0040
05:00:35 INFO | [qemu output] X02=000000000000ff80 X03=0000ffff909a0000 X04=0000000000000000
05:00:35 INFO | [qemu output] X05=ffffa072a22d0000 X06=ffffa072a22c0000 X07=cdcdcdcdcdcdcdcd
05:00:35 INFO | [qemu output] X08=cdcdcdcdcdcdcdcd X09=cdcdcdcdcdcdcdcd X10=cdcdcdcdcdcdcdcd
05:00:35 INFO | [qemu output] X11=cdcdcdcdcdcdcdcd X12=cdcdcdcdcdcdcdcd X13=cdcdcdcdcdcdcdcd
05:00:35 INFO | [qemu output] X14=cdcdcdcdcdcdcdcd X15=00007889c9852732 X16=ffff5eedb60cae20
05:00:35 INFO | [qemu output] X17=0000ffff91a81c60 X18=0000000000000001 X19=0000000000000000
05:00:35 INFO | [qemu output] X20=0000000000020000 X21=ffff00001cccfd40 X22=ffffa072a22d0000
05:00:35 INFO | [qemu output] X23=0000000000010000 X24=ffff7fe81ca88b00 X25=ffff00001cccfd50
05:00:35 INFO | [qemu output] X26=0000000110320000 X27=0000000000000000 X28=0000000000011032
05:00:35 INFO | [qemu output] X29=ffff00001cccfa70 X30=ffff5eedb60ca7e4  SP=ffff00001cccfa70
05:00:35 INFO | [qemu output] PSTATE=20000005 --C- EL1h
```

After this fix
xfs pass
```
# python3 ConfigTest.py --guestname=RHEL.8.4.0 --platform=aarch64 --driveformat=virtio_scsi --nicmodel=virtio_net --testcase=virtio_fs_share_data.run_stress.with_xfstest.with_local_source.with_cache.auto --clone=no
[snip]
INFO | Start to run test
JOB ID     : 124bd4fe9166871711c95c5cbdb84a70ee2b1d82
JOB LOG    : /root/avocado/job-results/job-2021-04-19T21.20-124bd4f/job.log
 (1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_xfstest.with_local_source.with_cache.auto.arm64-pci: PASS (2846.22 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-04-19T21.20-124bd4f/results.html
JOB TIME   : 2847.47 s
```

nfs failed: generic/123 regression
```
# python3 ConfigTest.py --guestname=RHEL.8.4.0 --platform=aarch64 --driveformat=virtio_scsi --nicmodel=virtio_net --testcase=virtio_fs_share_data.run_stress.with_xfstest.with_nfs_source.with_cache.auto --clone=no
[snip]
INFO | Start to run test
JOB ID     : 233e0af6e8207df9e559a9d76230f1c424b51879
JOB LOG    : /root/avocado/job-results/job-2021-04-19T22.29-233e0af/job.log
 (1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_xfstest.with_nfs_source.with_cache.auto.arm64-pci: -^[^[^FAIL: The xfstest failed. (3823.12 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-04-19T22.29-233e0af/results.html
JOB TIME   : 3824.37 s

```

liuyd96 commented:
@zhenyzha @YongxueHong . Please help me review it. Thanks !

zhenyzha commented on qemu/tests/cfg/virtio_fs_share_data.cfg:
```diff
@@ -146,7 +146,14 @@
                     mem = 32768
                     size_mem1 = 32G
-                    blacklist = generic/551
+                    tmpfs_size = 32G
```
Actually, the size of tmpfs depends on the size of host mem by default. In linux, it defaults to half of mem.
Like this:
```
[root@ampere-hr350a-04 ~]# df -h | grep /dev/shm | gawk '{ print $2 }'
128G
[root@ampere-hr350a-04 ~]# free -h
              total        used        free      shared  buff/cache   available
Mem:          254Gi       3.8Gi       240Gi        32Mi        11Gi       228Gi
Swap:         4.0Gi          0B       4.0Gi
```
So I think the bug error is that 'tmpfs' is too small because the host mem is less than 64G.
I think we should add a judgment to the code, when the host mem is less than 64G, manually set the tmpfs size.

zhenyzha replied:
@YongxueHong what do you think?

liuyd96 replied:
>  Actually, the size of tmpfs depends on the size of host mem by default. In linux, it defaults to half of mem.

Yes.
>  So I think the bug error is that 'tmpfs' is too small

Yes, 16G is not enough for this xfstest

>  when the host mem is less than 64G, manually set the tmpfs size.

Ok, i got it. If we need add a judgement why not check /dev/shm size directly 😃

zhenyzha replied:
Oh yes, you're right.
But I'm not sure, if host `/dev/shm` too small, we set it to 32G, when the test is done, do we need to restore its original size?
@YongxueHong

liuyd96 replied:
>  But I'm not sure, if host /dev/shm too small, we set it to 32G, when the test is done, do we need to restore its original size?
It already has the restore step in virtio_fs_share_data.py

zhenyzha replied:
Yes, I mean if we don't need to recover, we can omit this step.

zhenyzha replied:
After discussion with @YongxueHong
Since `mem-path_mem1 = /dev/shm` line 23, there is no need to add additional parameters, just use `size_mem1` as the minimum value.
On the other hand, the size needs to be restored after the test is done.

liuyd96 replied:
ok, got it. Thanks a lot.

liuyd96 reviewed (commented):
ok, got it.

zhenyzha commented on qemu/tests/virtio_fs_share_data.py:
```diff
@@ -233,6 +236,11 @@ def get_viofs_exe(session):
                 session.cmd(cmd_pjdfstest % fs_dest, io_timeout)
 
             if cmd_xfstest:
+                # Increase tmpfs /dev/shm to 32G
```
And since here is the method of directly converting to mem size without viewing the original dir size.
So it’s better to do it in the preprocess stage, like 107 line;
```
    if cmd_xfstest:
        ori_tmpfs_size = process.run(cmd_get_tmpfs, shell=True).stdout_text.replace("\n", "")
        logging.debug("original tmpfs size is %s", ori_tmpfs_size)
        if process.run(cmd_set_tmpfs % size_mem1, shell=True).exit_status:
            test.fail('xfstest set tmpfs failed.')
        if setup_local_nfs:
            for fs in params.objects("filesystems"):
                nfs_params = params.object_params(fs)
                params["export_dir"] = nfs_params.get("export_dir")
                params["nfs_mount_src"] = nfs_params.get("nfs_mount_src")
                params["nfs_mount_dir"] = nfs_params.get("fs_source_dir")
                nfs_local = nfs.Nfs(params)
                nfs_local.setup()
        params["start_vm"] = "yes"
        env_process.preprocess_vm(test, params, env, params["main_vm"])
```
@YongxueHong what do you think?

YongxueHong replied:
> And since here is the method of directly converting to mem size without viewing the original dir size.
> So it’s better to do it in the preprocess stage, like 107 line;
> 
> ```
>     if cmd_xfstest:
>         ori_tmpfs_size = process.run(cmd_get_tmpfs, shell=True).stdout_text.replace("\n", "")
>         logging.debug("original tmpfs size is %s", ori_tmpfs_size)
>         if process.run(cmd_set_tmpfs % size_mem1, shell=True).exit_status:
>             test.fail('xfstest set tmpfs failed.')
>         if setup_local_nfs:
>             for fs in params.objects("filesystems"):
>                 nfs_params = params.object_params(fs)
>                 params["export_dir"] = nfs_params.get("export_dir")
>                 params["nfs_mount_src"] = nfs_params.get("nfs_mount_src")
>                 params["nfs_mount_dir"] = nfs_params.get("fs_source_dir")
>                 nfs_local = nfs.Nfs(params)
>                 nfs_local.setup()
>         params["start_vm"] = "yes"
>         env_process.preprocess_vm(test, params, env, params["main_vm"])
> ```
> 
> @YongxueHong what do you think?

If you don't check the current size of tmpfs but set tmpfs size directly, suggest you could put them into `pre_command` and `post_command` in the cfg to do that.

YongxueHong replied:
Hi @zhencliu
Sorry, I wanna @zhenyzha

zhencliu replied:
Looks like this comment is for @zhenyzha . 

Agree, testing env setup in pre_command and cleanup in post_command are better(e.g. with some shell cmds), let's focus on the test assertions.

liuyd96 replied:
Ok, got it. Thanks for your memtion.
Since some testing env setup codes already exists, i will move them to pre/post_command in a new commit.

YongxueHong replied:
Anyway, your solution looks good to me. Up to you.

liuyd96 commented:
Hi, @zhenyzha 
I updated this PR and added one new commit  8740c2e
Based on above comments. I moved guest testing evn setup codes to guest_pre_command.

The other 2 commit are not changed.
I kept the cmd_get_tmpfs and cmd_set_tmpfs since they are 2 useful api.
We may use them in both setup and cleanup.

YongxueHong commented on qemu/tests/virtio_fs_share_data.py:
```diff
@@ -91,26 +91,32 @@ def get_viofs_exe(session):
+    if cmd_xfstest:
+        # /dev/shm is the default memory-backend-file, the default value is the
+        # half of the host memory. Increase it to guest memory size to avoid crash
+        ori_tmpfs_size = process.run(cmd_get_tmpfs, shell=True).stdout_text.replace("\n", "")
+        logging.debug("original tmpfs size is %s", ori_tmpfs_size)
+        if process.run(cmd_set_tmpfs % size_mem1, shell=True).exit_status:
+            test.fail('xfstest set tmpfs failed.')
```
```suggestion
        ori_tmpfs_size = process.run(cmd_get_tmpfs, shell=True).stdout_text.replace("\n", "")
        params["post_command"] = cmd_set_tmpfs % ori_tmpfs_size
        params["pre_command"] = cmd_set_tmpfs % size_mem1
```

We mean that set tmpfs size and recover tmpfs into pre-process and post-process like above.

zhenyzha commented on qemu/tests/cfg/virtio_fs_share_data.cfg:
```diff
@@ -146,7 +146,11 @@
                     image_snapshot = yes
                     mem = 32768
                     size_mem1 = 32G
```
pre_command = 'mount -o remount,size=${size_mem1} /dev/shm'
I think adding here is more in line with expectations

zhenyzha replied:
I overlooked the issue, ignore this suggestion, please

liuyd96 commented:
Verified on nfs test
```
# python3 ConfigTest.py --guestname=RHEL.8.4.0 --platform=aarch64 --driveformat=virtio_scsi --nicmodel=virtio_net --testcase=virtio_fs_share_data.run_stress.with_xfstest.with_nfs_source  --clone=no
JOB ID     : 9a103cd75682dd1f315cd9ff7c0cecab5f1f3041
JOB LOG    : /root/avocado/job-results/job-2021-04-23T02.29-9a103cd/job.log
 (1/1) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_xfstest.with_nfs_source.with_cache.auto.arm64-pci: PASS (3683.53 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-04-23T02.29-9a103cd/results.html
JOB TIME   : 3684.46 s
```

zhenyzha commented:
```
 (1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_xfstest.with_local_source.with_cache.auto.arm64-pci: PASS (5705.20 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.run_stress.with_xfstest.with_nfs_source.with_cache.auto.arm64-pci: PASS (8479.07 s)
```


```
 (1/3) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.arm64-pci: PASS (205.37 s)
 (2/3) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.with_reboot.arm64-pci: PASS (265.99 s)
 (3/3) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.aarch64.io-github-autotest-qemu.virtio_fs_share_data.with_cache.auto.xattr_on.flock_on.lock_posix_off.arm64-pci: PASS (182.33 s)

```

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

YongxueHong commented:
Hi, The first commit header is a bit long, please shorten it into 69 characters.

YongxueHong commented:
Please fix the CI check errors.Thanks.

liuyd96 commented:
Hi, @YongxueHong 
Looks like the CI errors are not related to this commit.
It's a inspekt params error
```
0.54s$ inspekt checkall --disable-style E501,E265,W601,E402,E722,E741 --enable-lint W0611,W1201 --no-license-check
inspekt: 'checkall --disable-style E501,E265,W601,E402,E722,E741 --enable-lint W0611,W1201 --no-license-check' is not an inspekt command. See 'inspekt --help'.
Did you mean one of these?
  license
The command "inspekt checkall --disable-style E501,E265,W601,E402,E722,E741 --enable-lint W0611,W1201 --no-license-check" exited with 2.
```

YongxueHong commented:
Yes, you are right. Maybe there is a workaround you could have a try:
git checkout master
git fetch --all
git pull [upstream branch] master
git checkout [develop branch]
git rebase master
git pull -f origin [develop branch]

YongxueHong commented:
Hi @liuyd96 
Looks like the workaround doesn't work.
So let's merge it.
